File: demos.m
%% newest workable code
rand('seed',54321);
randn('seed',54321);

%% parameter for HMC
D = 2;
k1=99.;
SIGMA=[1,k1/100.;k1/100.,1];
%%SIGMA=[k1,1/k1;1/k1,k1];
ISIGMA = inv(SIGMA);
U = @(x) x' * ISIGMA * x/2;
dU = @(x) ISIGMA *x;
x0 = zeros(D,1);
if 1
[x, p] = ghmc(U,dU, D, x0, EPISODE=10000);
else
  [x, p] = hmc(U,dU, D, x0, EPISODE=10000,delta=.5);
end

cov(x(:,2000:end)')
figure;
plot(x(1,2000:end),x(2,2000:end),'.');
axis equal;
hold on;
t = linspace(0,2*pi,100)';
xy = 3*sqrtm(SIGMA)*[cos(t)';sin(t)'];
plot(xy(2,:),xy(1,:),'-','Linewidth',2);

pause;

File: demos.py
import numpy as np
import matplotlib.pyplot as plt
from ghmc import ghmc

np.random.seed(54321)

D = 2
k1 = 70.
SIGMA = np.array([[1,k1/100.],[k1/100.,1]])
##SIGMA = np.array([[k1,1/k1],[1/k1,k1]])
U = lambda x: np.dot(x, np.dot(np.linalg.inv(SIGMA), x)) / 2
dU = lambda x: np.dot(np.linalg.inv(SIGMA),x)
if True:
    info = ghmc(U, dU, D,EPISODE=10000)
else:
    info = hmc(U, dU, D,EPISODE=1000, VERBOSE=True, delta = 0.5)

x = info['x']

print(np.cov(np.transpose(x[2000:,:])))

plt.plot(x[2000:,0],x[2000:,1],'.')
plt.show()
plt.pause(10)
